### PR TITLE
Use react/async for promise wait

### LINKED ISSUE #49: Use fibers instead of stopping the event loop waiting for a response to complete
**Description**
Currently, the promise object of this package contains a major non-blocking design flaw that stops the loop while waiting for a response to complete: https://github.com/php-http/react-adapter/blob/master/src/Promise.php#L115-L132

**Example**
When using the upcoming fiber implementation in [`react/async`](https://github.com/reactphp/async) this design flaw can be turned into a feature and still wait for responses to come in and also keep the event loop running when it does.

**Additional context**
The `v4` version (already usable at `^4@dev`) requires PHP 8.1. However we're also preparing a `v2` and `v3` that provide the same `await()` method but by using the same technique as the `wait` method on the promise object as an upgrade path for those that can't upgrade to PHP 8.1 yet.

Would you be open to a PR that implements this? Either directly to 8.1 or with keeping support for older PHP version by using the `v2` and `v3` packages as well.

P.S. I understand that calling the `wait` method a design flaw might come over as harsh. But this comes from stopping the loop in a fully non-blocking application is a deadly sin as the application will randomly stop as nothing is expecting this. I also fully understand that this is fine in hybrid applications this is fine or even expected, hence we added `v2` and `v3` as upgrade path for this situations.

### PR DESCRIPTION
**Pull request by WyriHaximus**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | yes
| Related tickets | fixes #49
| Documentation   | if this is a new feature, link to pull request in https://github.com/php-http/documentation that adds relevant documentation
| License         | MIT

#### What's in this PR?

Change promise implementation to not block

#### Why?

Avoid blocking, see #49

#### Checklist

- [ ] Updated CHANGELOG.md to describe BC breaks / deprecations | new feature | bugfix
- [ ] Documentation pull request created (if not simply a bugfix)

### CODE REVIEW

**Comment by dbu** [comment]
please let me know if you need any input (i have no experience with react though)

**Comment by WyriHaximus** [comment]
@dbu Will do! I think the only thing might be the `@dev` targeting in `composer.json`, but lets get the build green first. Will also update the workflow to run 8.1 and drop the older version. Will ping you when I need something from your end.

**Comment by WyriHaximus** [comment]
@dbu Filed #51 and #52 to run tests against PHP 8.1, one is a simple addition of `8.1`, while the other finds the supported PHP versions in range for the project. Take your pick :D

File: src/Promise.php [hunk 1]
```diff
@@ -114,12 +115,9 @@ public function getState()
...
     {
-        $loop = $this->loop;
-        while (HttpPromise::PENDING === $this->getState()) {
-            $loop->futureTick(function () use ($loop) {
-                $loop->stop();
-            });
-            $loop->run();
+        try {
+            await($this->promise);
+        } catch (\Throwable) {
```
Comment (dbu) [inline src/Promise.php:120]: is it correct to eat away the exception and not do anything with it? seems dangerous to me.

Comment (WyriHaximus) [reply src/Promise.php:120]: Check lines 82 to 96, and 123 to 129. In theory that will cover the swallowing here. This is something I'm considering changing but not sure what the best approach is yet

Comment (dbu) [reply src/Promise.php:120]: ah, i see yep. thanks for the explanation. 

what happens with line 128 if there was an exception and we chose not to unwrap? the response will be null, right? but that will have been the same before if i understand correctly.

**Comment by dbu** [comment]
i fixed the static analysis in the master branch. can you please rebase?

what is needed to wrap this MR up?

File: src/Promise.php (hunk 1 above)
Comment (WyriHaximus) [reply src/Promise.php:120]: Yes, I kept the behavior the same as before. Not unwrapping is a very conscious call to swallow any error that might come up.

**Comment by dbu** [comment]
> > what is needed to wrap this MR up?
> 
> A stable release of `react/async` which is the case since Monday

ah nice! i will now be away until next week, will check back then and tag a release of this adapter if its ready.

**Comment by WyriHaximus** [comment]
> i fixed the static analysis in the master branch. can you please rebase?

@dbu Ok so funny thing, it seems that the static analysis runs on PHP 8.0, and this PR raises the minimum PHP version to 8.1 :joy:

**Comment by dbu** [comment]
> @dbu Ok so funny thing, it seems that the static analysis runs on PHP 8.0, and this PR raises the minimum PHP version to 8.1 joy . Code coverage has a similar issue, but that is running on PHP 7.4.

can you please adjust the php version for code coverage? 

with static analysis its a bit more tricky, as we use the oskarstark phpstan image. i created https://github.com/OskarStark/phpstan-ga/pull/53 to try to upgrade to php 8.1

**Comment by WyriHaximus** [comment]
> > @dbu Ok so funny thing, it seems that the static analysis runs on PHP 8.0, and this PR raises the minimum PHP version to 8.1 joy . Code coverage has a similar issue, but that is running on PHP 7.4.
> 
> can you please adjust the php version for code coverage?

Oops I missed that was hardcoded. Just updated that.

> with static analysis its a bit more tricky, as we use the oskarstark phpstan image. i created [OskarStark/phpstan-ga#53](https://github.com/OskarStark/phpstan-ga/pull/53) to try to upgrade to php 8.1

How open are you to alternatives? Could look into using a github action for it.

**Comment by dbu** [comment]
> How open are you to alternatives? Could look into using a github action for it.

the thing we use is a github action ;-)

oskar already merged the upgrade and tagged it. i now restarted the build and its green \o/


the codestyle error is because it wants the use function to have a newline before and after - maybe put the `use function` after the list of classes we use?

**Comment by WyriHaximus** [comment]
> > How open are you to alternatives? Could look into using a github action for it.
> 
> the thing we use is a github action ;-)

Oh hah it looked like it was just a docker container.

> oskar already merged the upgrade and tagged it. i now restarted the build and its green \o/

Great!

> the codestyle error is because it wants the use function to have a newline before and after - maybe put the `use function` after the list of classes we use?

Tried that, but it wants the empty lines before and after it in the current position

**Comment by dbu** [comment]
sorry about the `use function` thing. not sure why they do it like that, its a bit silly. but i rather have it look a bit silly than open the can of worms of discussing about codestyle :smile: 

thanks a lot for the contribution, i will merge now. should this be a version 3 or does the change merit a major version bump? 

and should i already tag a release, or are you doing more testing?

**Comment by WyriHaximus** [comment]
> sorry about the `use function` thing. not sure why they do it like that, its a bit silly. but i rather have it look a bit silly than open the can of worms of discussing about codestyle smile

No worries, I respect and follow the CS rules of the project without complaining, how silly they might look to me :smiley: .

> thanks a lot for the contribution, i will merge now. should this be a version 3 or does the change merit a major version bump?

Would go for a major version bump IMHO because the mechanic of how you wait changes and now requires fibers.

> and should i already tag a release, or are you doing more testing?

Give me a week to test it in a project. Need to set up Sentry for it, should be fun :D

**Comment by dbu** [comment]
great! i merged and set up the 4.x branch for it (getting rid of confusing & politically incorrect `master` branch while i am at it)

can you please update the changelog with more details and maybe a short instruction how to adapt usage?

and let me know when to tag a release ;-)

**Comment by WyriHaximus** [comment]
> > and should i already tag a release, or are you doing more testing?
> 
> Give me a week to test it in a project. Need to set up Sentry for it, should be fun :D

That took a bit longer than intended sorry. But I come bearing the goodest of news

> great! i merged and set up the 4.x branch for it (getting rid of confusing & politically incorrect `master` branch while i am at it)

:+1:

> can you please update the changelog with more details and maybe a short instruction how to adapt usage?

Will do :+1:.
 
> and let me know when to tag a release ;-)

Let me file a changelog update and we're good to go. I've tested the 4.x branch in a project for Sentries HTTP transport. Running without any issues what so ever <3.

**Comment by WyriHaximus** [comment]
Ok PR for fibers explanation in the changelog is up at https://github.com/php-http/react-adapter/pull/57 and also filed https://github.com/php-http/react-adapter/pull/56 with some clean up

**Pull request merged**